browser_console.js

```diff
@@ -0,0 +1,159 @@
+            return {
+                __pentf_serialized: true,
+                type: 'Error',
+                message: value.message,
+                stack: value.stack,
+            };
+        } else if (value instanceof Set) {
+            return {
+                __pentf_serialized: true,
+                type: 'Set',
+                items: Array.from(value).map(item => serialize(item, seen)),
+            };
+        } else if (value instanceof Map) {
+            return {
+                __pentf_serialized: true,
+                type: 'Map',
+                items: Array.from(value.entries()).map(entry => {
+                    return [serialize(entry[0], seen), serialize(entry[1], seen)];
+                }),
+            };
+        }
+
+        seen.add(value);
+        let out = {};
+        Object.keys(value).forEach(key => {
+            out[key] = serialize(value[key], seen);
+        });
+
+        return out;
+    } else if (typeof value === 'function') {
+        return {
+            __pentf_serialized: true,
+            type: 'Function',
+            name: value.name,
+        };
+    }
+
+    return value;
+}
+
+/**
+ * Serialize console arguments and send them to puppeteer. Unfortunately for
+ * us the native serialization methods for JSHandle objects from puppeteer
+ * are lossy. They turn Error objects into `{}` and `null` to `undefined`.
+ * The passed "preview" object is incomplete and truncates all data for nested
+ * objects or other complex values.
+ *
+ * The only way to keep the data intact is to use a custom serialization format
+ * and pass it around as a string.
+ *
+ * @param {import('puppeteer').Page} page
+ */
+async function forwardBrowserConsole(page) {
+    // The stack is not present on the trace method, so we need to patch it in
+    await page.evaluate((fn) => {
+        const serialize = new Function(`return ${fn}`)();
+        const native = {};
+        native.trace = console.trace;
+        console.trace = (...args) => {
+            const stack = new Error().stack.split('\n').slice(2).join('\n');
+            args = ['\n' + stack, ...args];
+            native.trace.apply(null, args.map(arg => serialize(arg, new Set())));
+        };
+    }, serialize.toString());
+
+    page.on('console', async message => {
+        let type = message.type();
+        // Correct log type for warning messages
+        type = type === 'warning' ? 'warn' : type;
+
+        try {
+            const args = await Promise.all(
+                message.args().map(arg => {
+                    return arg.executionContext().evaluate((handle, fn) => {
+                        const serialize = new Function(`return ${fn}`)();
+                        return serialize(handle, new Set());
+                    }, arg, serialize.toString());
+                })
+            );
+
+            const parsed = args.map(arg => parseConsoleArg(arg));
+            if (type === 'trace') {
```

Comment:
It's not necessary for this PR, but in the future we may want to coordinate with the `output` module here, so that it stops its output before and resumes afterwards.

Reply:
Yeah, for that to work the logging function needs to support different log levels and variadic arguments. That's something better done in a separate PR 👍